### DISCUSSION TITLE
IfLetStore .onAppear or .task gotcha

### DISCUSSION THREAD

**Discussion by tgrapperon** [pointfreeco/swift-composable-architecture#804] [Q&A] [answered]
I've encountered a setup where `IfLetStore`'s were not behaving as I would expect in the first place. This is not a bug *per se*, but the documentation can maybe be extended to mention this configuration.

In my setup, I have a list of optional states, each corresponding to the value of some enum (the states need to stay alive once initialized, even if the case changes, so I'm not using `SwitchStore`, but it works in the same spirit). 

At the beginning, all the states are `nil` and some `onAppear` sends an action to initialize the correct state corresponding to the current enum's value. This non `nil` state then activate some `IfLetStore` and the view is populated. In the body of the view, I had something like:
```swift
Group {
  switch viewStore.kind {
    case .first: IfLetStore(store.scope…
    case .second: IfLetStore(store.scope…
  }
}
.onAppear { viewStore.send(.initialize) }
```
The `onAppear` method wasn't called. I understand why now, but I assumed it would be.
If some `else:` view is provided on any of the `IfLetStore`, `onAppear` is called as expected.

In other words, `Group` of `IfLetStore`'s are not triggering `.onAppear` or `.task` when all the states are `nil`. I think this is naturally expected for individual `IfLetStore`'s, but this can be surprising at first for `Group`s.

I don't know if it's worth adding a note in the documentation of `IfLetStore`. I didn't find a way to explain it concisely.

**Comment by mbrandonw** [comment, marked as answer]
Hey @tgrapperon, yeah this is definitely tricky, but it is indeed how SwiftUI and TCA is expected to work, and this exact thing would happen in vanilla SwiftUI.

It's tripped me up a number of times, but you have to think of view modifiers applied to `Group` as being applied to every view inside. So:

```swift
Group {
  IfLetStore(...)
  IfLetStore(...)
}
.onAppear { ... }
```

Is really:

```swift
IfLetStore(...)
  .onAppear { ... }
IfLetStore(...)
  .onAppear { ... }
```

Further, `IfLetStore` is really just:

```swift
if condition {
  content
}
```

so your original code further transforms into something like this:

```swift
if condition {
  content
    .onAppear { ... }
}
if condition {
  content
    .onAppear { ... }
}
```

And so from this it becomes more clear that `.onAppear` is only going to be called when something inside the `IfLetStore` is actually shown.

If you really need to kick off the initialization work without there being a view visible on the screen you might have to resort to some hacky things, such as `.background(...)` with a hidden view that has the `.onAppear`.

**Reply by tgrapperon** [reply]
I think it works the same for `ForEach`. I've just looked at `Group` documentation, and they mention this case exactly. I've should have checked this before opening the issue.
I've solved my problem by initializing some substate manually if needed when it is derived from the root state.
I think replacing the `Group` by a `ZStack` would have been enough otherwise.
Thanks for the response!

**Comment by mbrandonw** [comment]
Also I'm going to convert this to a discussion since it's not really a bug with the library.
